### PR TITLE
Version 0.1

### PR DESCRIPTION
**Pull request by mecha**
Initial release.

Contains solutions for #2 and #3.

### CODE REVIEW

**Review by XedinUnknown — COMMENTED** [review]
How about adding a Composer `build` script? This would save typing the path to `vendor/bin/phing`.

**Comment by XedinUnknown** [comment]
After creation finished, there was no `master` branch. I would much prefer there to be both the `develop` and `master`, and to be on `master` after creation.

**Comment by XedinUnknown** [comment]
I think `build.xml` needs to be removed after creation, too.

**Comment by mecha** [comment]
@XedinUnknown Unfortunately, due to a [Composer bug](https://github.com/composer/composer/issues/3299) the build script would not stop at the prompts.

There should be both a `master` and `develop` branch - but even so I agree to be on `master` by default. I will confirm the absence of `master` and make it default if not already so.

**Comment by XedinUnknown** [comment]
> Unfortunately, due to a Composer bug the build script would not stop at the prompts.

@mecha, fair enough. Maybe add this in a "Known Issues" section? Or, just create an issue.

**Comment by XedinUnknown** [comment]
@mecha, when you fix the absence of `master`, and make it the default, as well as remove the `build.xml` file, maybe also add all files to index?

**Comment by mecha** [comment]
@XedinUnknown That's a good idea. Maybe even make an "Initial commit"?

**Comment by XedinUnknown** [comment]
Maybe the commit is too much, since we may wanna add some other config files.

**Comment by XedinUnknown** [comment]
Single slashes didn't get converted to double slashes for me, and also the slashes weren't added at the end. I entered `Dhii\Wp\I18n`.

**Comment by XedinUnknown** [comment]
Another thing that could be helpful:
`$ git remote add origin https://github.com/dhii/new-project`

**Comment by XedinUnknown** [comment]
@mecha, looks like the composer deps didn't get installed at project creation time. That's why it's so fast.

**Comment by mecha** [comment]
@XedinUnknown Test now. All issues should be addressed.

**Comment by mecha** [comment]
An FYI to all testers, the command for testing is:

```
composer create-project dhii/project-bootstrap:dev-release/0.1 my-project
```

After releasing, the version constraint can be eliminated.

**Comment by mecha** [comment]
@XedinUnknown Issue #3 has been addressed and #2 has been implemented in the build process.

**Comment by XedinUnknown** [comment]
Working very well. No more suggestions or required questions. I think this is ready for merge. Will just try it on a couple of other cases, just to make sure.

**Comment by XedinUnknown** [comment]
Add `composer phing` ability with `scripts`, if possible, please, @mecha.

**Comment by mecha** [comment]
@XedinUnknown Not possible. See [this comment](https://github.com/Dhii/bootstrap/pull/1#issuecomment-286836923) further up.

What I _will_ do is add test templates in the test sub-directories, as agreed in an earlier conversation.

**Comment by XedinUnknown** [comment]
When I write "Dhii" for vendor, the human readables e.g. README and License should show it as written, i.e. "Dhii", but Composer, NB proj, and such, should see the slug, i.e. "dhii". Right now, it didn't happen, and NB project name, as well as the Composer package name, had "Dhii", i.e. verbatim.

**Comment by mecha** [comment]
Ah I see. You entered "Dhii" for the _vendor_ name. Right now, the script expects you to enter the vendor name correctly and does not do any manipulations to it. What you want to do is enter "Dhii" as the _author_.

**Comment by XedinUnknown** [comment]
@mecha, yes, I enter "Dhii" or "Dhii Team" or whatever I want the author to be as the author, and that goes correctly. The vendor name used in configuration etc, versus what is used for humans, must be sanitized, e.g. lower- and kebab-cased. The default of that should be derived from the author.

**Comment by XedinUnknown** [comment]
The dummy test's `createInstance()` method mocks `methods()` method. I think this should not be there.

**Comment by XedinUnknown** [comment]
Also, the FQNs used in dummy tests contain stuff like `_vendor\_package\FuncTest\_ns`. Don't we already know the root namespace to substitute `_vendor\_package` with? The `_ns` looks extra here, as the first classes are likely to be right in the root NS + test NS.

**Comment by XedinUnknown** [comment]
Another useful feature could be to collapse dashes when converting HR -> slug. Example: "Data - Tree - Abstract" becomes `data-tree-abstract` instead of `data---tree---abstract`. Should be very easy to achieve: convert any number of symbols from set, such as `[ _-+\/]`, into one symbol.

**Comment by XedinUnknown** [comment]
Using the [`prefer-stable`] flag could really improve bootstrap speed, as it would cause cached versions to be used for many packages.

[`prefer-stable`]:                          https://getcomposer.org/doc/04-schema.md#prefer-stable

**Comment by mecha** [comment]
@XedinUnknown Addressed the issues.

File: build-functions.php [hunk 1]
```diff
@@ -29,7 +29,7 @@ function kebabCaseToUcWords($kebab)
  */
 function toKebabCase($string)
 {
-    return implode('-', explode(' ', strtolower($string)));
+    return preg_replace('/[ _\\-\\+\\/\\\\]+/', '-', strtolower($string));
```
Comment (XedinUnknown) [inline build-functions.php]: This is wrong. See [example](http://www.phpliveregex.com/p/kCA). You should probably first replace all separation chars, e.g. `- +_/\\`, with `-`, and then collapse all collapsible chars similarly to what you did, but capture the actual char, and replace with a single instance of that char, e.g. `$0`.

Comment (mecha) [reply build-functions.php]: Didn't understand the second part, after "and then collapse ...".

But note that the example on the site needs to be de-escaped. [See now](http://www.phpliveregex.com/p/kCB).

Comment (mecha) [reply build-functions.php]: Amended the function. Now uses a `collapseChars()` function.

**Comment by XedinUnknown** [comment]
@mecha, I think it may also be worth adding `prefer-stable` to this repo's Composer config as well.

**Comment by mecha** [comment]
@XedinUnknown Agree.

**Comment by mecha** [comment]
@XedinUnknown Perhaps also use `master` for `dhii/php-cs-fixer-config`?

**Comment by XedinUnknown** [comment]
I don't think ready yet. We'd have to bump up the required PHP version for that. And we still need PHP 5.3 for many of the interfaces.

**Comment by mecha** [comment]
@XedinUnknown Added the `prefer-stable` option in the composer config template.

Any last changes? I'd like to merge this and release.

**Comment by XedinUnknown** [comment]
OMG this merge was a mistake. @mecha, I wrote to add
> `prefer-stable` to *this* repo's Composer config as well

I have no idea how you added it to the template, especially since I already did that earlier, in https://github.com/Dhii/bootstrap/commit/13c451de75227efea4e1208b079504ecd106e1e8.

**Comment by mecha** [comment]
No, _I_ added it to the `composer.json` template. You added it to the `composer.json` of the bootstrap.

**Comment by XedinUnknown** [comment]
You didn't click the commit link?

**Pull request merged**